### ISSUE TITLE
A way to check what is queued or avoiding duplicates

### ISSUE THREAD

**Issue by patrikhermansson** [SGrondin/bottleneck#29]
Hi. I use bottleneck to handle API requests that have a rather low limit, therefore my queue can go upwards to 1000-2000 items. The problem I have is that some new tasks will in some circumstances already be in the queue but not completed. Is there a way to check what is in the queue and the arguments, in order to avoid duplicates?

**Comment by SGrondin** [comment]
This is not currently possible out of the box, but you could implement it in your code by keeping track of what has been added in an Object `{}` and deleting keys once the request has completed.

However, it would be an interesting feature to add to Bottleneck! In your use case, do you have a `requestId` that Bottleneck could use to determine whether the request has already been added? Or do you rely on the arguments to know if it's already been added?

**Comment by patrikhermansson** [comment]
Yes. That's what I am doing right now! 

I'm dealing with a twitter stream and need to perform some tasks on the users I get. I don't need to do those tasks multiple times per user, so I use the twitter user name as an identifier and add that to a simple array that I check towards before scheduling new tasks with bottleneck. Then I remove them when task is complete.

It works well but becomes problematic there is too much activity in the stream and queue reaches its limit and tasks get dropped. It's possible but requires a lot of work to basically emulate the behaviour of the highwater strategy. And just using an array for keeping the list is a long term memory leak if there are tasks that get dropped.

I tried to solve this by using the 'dropped' event. But that does not give me the arguments of the dropped task.

**Comment by SGrondin** [comment]
There's a bug in the `dropped` event. I'm fixing it now and I'll release a fix shortly.

**Comment by SGrondin** [comment]
I just released `v1.15.1` with a fix to propagate the arguments into the task runner so you get them in the `dropped` event. That way you'll be able to use `dropped` to remove requestIDs from your tracker. And by the way, I really recommend you use an object (`{}`) to keep track if those usernames instead of an array. Delete from the object with `delete trackerObject[username];`.

Please let me know if that doesn't fix your issue.

Have a look at the [`rejectOnDrop`](https://github.com/SGrondin/bottleneck#constructor) option in case it would be useful to you.

**Comment by SGrondin** [comment]
I'm still open to the idea of adding this sort of tracker to Bottleneck itself, but I think it requires careful thought.

Pros:
- Nice to have. It's the kind of details that Bottleneck was built to handle.

Cons:
- What would it look like from the client's perspective? There's already 4 ways to add a task to Bottleneck: `submit(task, args..., cb)`, `submitPriority(priority, task, args..., cb)`, `schedule(promiseProvider, args...)` and `schedulePriority(priority, promiseProvider, args...)`. Maybe Bottleneck needs a generic method that takes an object argument to pass all those parameters? But that breaks the flow of user-side code.
- It means implementing the same kind of Garbage Collection logic that Bottleneck.Cluster uses to expire unused limiters.

Thoughts? My main concern is how to keep Bottleneck clean and easy to use from the user's perspective.

---

@patrikhermansson Here's a crazy idea for you:
You could entirely implement that tracker by using a Cluster to manage the keys, setting each generated limiter's `reservoir` to 1 and `chain`ing it to a global limiter that would handle the real rate limits. It's crazy, but it would work 100% of the time and Bottleneck would handle all the logic for you. I wouldn't recommend it for 20,000,000 requests, but you're only dealing with 2,000 so it's not an issue at all.

I was curious what it would look like, so I implemented it. It was a lot easier than I thought and it worked on the first try. It manages all the tracking for you. I commented it so it would be easy to follow. Run it and the output will show you that it's indeed doing the right thing.

```js
var Bottleneck = require('bottleneck')

// Create a cluster to manage the usernames.
// The cluster's rate limits don't matter, so we leave them empty.
var cluster = new Bottleneck.Cluster()

// Create a limiter to manage the real rate limits.
// That's where you put your real rate limits,
// let's say they are 1 and 1000.
var limiter = new Bottleneck(1, 1000)

// Here are your items to process.
// In this example, each item is a list of numbers to sum up, and a username.
// As you can see, `userA` is there twice.
// Only the first `userA` will be processed.
// So the results should be 6 (1+2+3), 7 (5+2) and 9 (3+3+3).
var items = [
  { username: 'userA', data: [1,2,3] },
  { username: 'userB', data: [5, 2] },
  { username: 'userA', data: [8] },
  { username: 'userC', data: [3, 3, 3] }
]

var processData = function (numbers) {
  // This is the work that you have to do to process each item.
  // Here all we do is sum up each array.
  // In your real application this is the heavy processing happens.
  console.log('Processing data:', numbers)
  return Promise.resolve(numbers.reduce((a, b) => a + b))
}

var addToBottleneck = function (item) {
  console.log('Adding', item, 'to Bottleneck!')
  var localLimiter = cluster.key(item.username)

  if (localLimiter.reservoir === null) {
    // This is a new limiter
    localLimiter.changeReservoir(1)
    localLimiter.chain(limiter)
  }

  localLimiter.schedule(processData, item.data).then(console.log)
}

items.forEach(addToBottleneck)
```

**Comment by patrikhermansson** [comment]
Thank you very much for your quick fix and helpful comments. I've been away and haven't been able to work on this yet. That's a really interesting solution actually! Right now I'm trying just to use the args from the 'drop' event, it should be enough I think

About implementing a queue: I don't really see the need for more parameters in the submit/schedule functions. The way I imagine this would be used is in cases similar to mine, where the data you bass into Bottleneck always differs and should be unique. So I just picture a function that you call on the limiter. E.g: limiter.tasklist(priority?) -> Returns array of objects with the arguments that was passed into the function. Then it's quite flexible and different implementations will use different arguments to check if a new task should be added to the queue. I would keep the interface as clean as it is now.

**Comment by SGrondin** [comment]
The problem with keeping track of arguments _forever_ is that Bottleneck is used in many long-running applications and this would simply take up too much memory over time.

I'm glad you like that solution. Something to keep in mind is that, as described in the docs, the Cluster will get rid of limiters that haven't been used within the last 5 minutes. In your case, it means that the same `username` would go through if it was added more than 5 minutes after the previous occurence of the same `username` was added. Calling `cluster.stopAutoCleanup()` disables that behaviour and you wouldn't have to worry about that edge case. I edited the example above to add `cluster.stopAutoCleanup()` to it.

What I've decided to do is to add a `changeTimeout()` method to Cluster to make it possible to adjust the `5 minutes` figure.

I think that this addition and the `1.15.1` fix are enough to address your use case. I'll close this Issue once this new feature has been released. Thoughts?
